File: algorithm/c++/MusicProgram2623/musicProgram2623.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

int N,M;
vector<int> adj[1001];
int degree[1001];
queue<int> que;

int main()
{
    ios_base::sync_with_stdio(0);
    cin.tie(0), cout.tie(0);
    
    cin >> N >> M;

    for (int i = 0; i < M; i++)
    {
        int cnt;
        cin >> cnt;

        vector<int> inputs_v;

        for (int j = 0; j < cnt; j++)
        {
            int input;
            cin >> input;
            inputs_v.push_back(input);   
        }
        
        for (int j = 0; j < cnt-1; j++)
        {
            adj[inputs_v[j]].push_back(inputs_v[j+1]);
            degree[inputs_v[j+1]]++;
        }
    }

    for (int i = 1; i <= N; i++) 
    {
        if (degree[i] == 0) que.push(i);
    }

    vector<int> result_vec; 

    while (!que.empty())
    {
        int cur = que.front();
        que.pop();
        result_vec.push_back(cur);

        for (int i = 0; i < adj[cur].size(); i++)
        {
            int next = adj[cur][i];
            degree[next]--;

            if (degree[next] == 0) que.push(next);
        }
    }

    if (result_vec.size() != N) cout << 0 << endl;
    else
    {
        for (int i = 0; i < result_vec.size(); i++) cout << result_vec[i] << endl;
    } 


    return 0;
}

File: algorithm/python/MusicProgram2623/musicProgram2623.py
import sys
from queue import Queue

input = sys.stdin.readline

N, M = map(int,input().split())

adj = [[] for _ in range(N+1)]
degree = [0] * (N+1)
queue = Queue()

for i in range(0,M):
    
    data = list(map(int, input().split()))
    input_len = data[0]
    input_list = []

    for j in range(1,input_len + 1):
        input_list.append(data[j])

    for j in range(0,len(input_list)-1):
        adj[input_list[j]].append(input_list[j+1])
        degree[input_list[j+1]] = degree[input_list[j+1]] + 1

for i in range(1,N+1):
    if (degree[i] == 0):
        queue.put(i)

result_list = []

while not queue.empty():
    cur = queue.get()
    result_list.append(cur)

    for i in range(0,len(adj[cur])):
        next = adj[cur][i]
        degree[next] = degree[next] - 1

        if (degree[next] == 0):
            queue.put(next)


if (len(result_list) == N):
    for i in range(0,len(result_list)):
        print(result_list[i])
else:
    print(0)


